### ISSUE TITLE
Note for 《Deeplearning with Pytorch》

### ISSUE THREAD

**Issue by cq19990718** [cq19990718/blog#19]
## Hadamard积
两个矩阵的按元素乘法称为Hadamard积（Hadamardproduct）（数学符号⊙）
A⊙B  用python表示为
```python
A * B
```
## 降维
求和
python中调用计算求和的函数：`.sum()`
默认情况下，调用求和函数会沿所有的轴降低张量的维度，使它变为一个标量。我们还可以指定张量沿哪一
个轴来通过求和降低维度。以矩阵为例，为了通过求和所有行的元素来降维（轴0）：`.sum(axis=0)`
指定axis=1将通过汇总所有列的元素降维：`.sum(axis=1)`
沿着行和列对矩阵求和，等价于对矩阵的所有元素进行求和：
```python
.sum(axis=[0, 1])
```
还有求均值等
## 点积
点积（dotproduct）x⊤y（或⟨x,y⟩）是相同位置的按元素乘积的和
我们可以通过执行按元素乘法，然后进行求和来表示两个向量的点积：
```python
torch.sum(x * y)
torch.dot(x, y)
```
## 矩阵-向量积
当我们为矩阵A和向量x调用torch.mv(A,x)时，会执
行矩阵‐向量积。注意，A的列维数（沿轴1的长度）必须与x的维数（其长度）相同。
```python
torch.mv(A,x)
```
## 矩阵-矩阵乘法
在A和B上执行矩阵乘法。这里的A是一个5行4列的矩阵，B是一个4行3列的矩阵。两者
相乘后，我们得到了一个5行3列的矩阵。
```python
torch.mm(A,B)
```
## 范数
在线性代数中，向量范数是将向量映射到标量的函数f。给定任意向量x，向量范数要满足一些属性。第一个
性质是：如果我们按常数因子α缩放向量的所有元素，其范数也会按相同常数因子的绝对值缩放：`f(αx)=|α|f(x).`
第二个性质是熟悉的三角不等式: `f(x+y)≤f(x)+f(y).`
第三个性质简单地说范数必须是非负的:` f(x)≥0.`
L2范数:
```python
 torch.norm(u)
```
深度学习中更经常地使用L2范数的平方
L1范数：
```python
torch.abs(u).sum()
```
Frobenius范数:
![image](https://github.com/user-attachments/assets/adcc499d-caa5-40b9-afdf-af5fbf12c461)
```python
torch.norm(torch.ones((4, 9)))
```
输出为6
## 梯度
![image](https://github.com/user-attachments/assets/08543941-3d73-4083-ae3b-65075cd1e9f7)
![image](https://github.com/user-attachments/assets/5acfd74e-d11b-4587-84bd-6f346a1f83e0)
一个简单的例子
```python
 x = torch.arange(4.0)       # output   tensor([0., 1., 2., 3.])
```
在我们计算y关于x的梯度之前，需要一个地方来存储梯度，注意，一个标量函数关于向量x的梯度是向量，并且与x具有相同的形状。
```python
x.requires_grad_(True) # 等价于x=torch.arange(4.0,requires_grad=True)
 x.grad # 默认值是None
```
函数y=2x⊤x
```python
y = 2 * torch.dot(x, x) 
```
通过调用反向传播函数来自动计算y关于x每个分量的梯度，并打印这些梯度。
```python
y.backward()
 x.grad    # output  tensor([ 0., 4., 8., 12.])
```
函数y=2x⊤x关于x的梯度应为4x
计算另一个函数
```python
# 在默认情况下，PyTorch会累积梯度，我们需要清除之前的值
x.grad.zero_()
 y = x.sum()
 y.backward()
 x.grad   # output tensor([1., 1., 1., 1.])
```
## 分离计算
这里可以分离y来返回一个新变量u，该变量与y具有相同的值，但丢弃计算图中如何计算y的任何信息。换句
话说，梯度不会向后流经u到x。因此，下面的反向传播函数计算z=u*x关于x的偏导数，同时将u作为常数处理，
而不是z=x*x*x关于x的偏导数。
```python
x.grad.zero_()
 y = x * x
 u = y.detach()
z = u * x
 z.sum().backward()
 x.grad == u    output  tensor([True, True, True, True])
```

## 查阅资料
调用dir函数，我们可以查询随机数生成模块中的所有属性：
```python
dir() 
```
通常可以忽略以“__”（双下划线）开始和结束的函数，它们是Python中的特殊对象，或以单个“_”（单下划
线）开始的函数，它们通常是内部函数。
有关如何使用给定函数或类的更具体说明，可以调用help函数。
```python
help() 
```
在Jupyter记事本中，我们可以使用?指令在另一个浏览器窗口中显示文档。例如，`list?`指令将创建
与`help(list)`指令几乎相同的内容，并在新的浏览器窗口中显示它。此外，如果我们使用两个问号，如`list??`，
将显示实现该函数的Python代码。
## 矢量化加速
我们需要对计算进行矢量化，从而利用线性代数库，而不是在Python中编写开销高昂的for循环。
矢量化代码通常会带来数量级的加速。另外，我们将更多的数学运算放到库中，而无须自己编写那么多的计算，从而减少了出错的可能性。
## 正态分布与平方损失
这是正态分布
![image](https://github.com/user-attachments/assets/b1e40e85-c227-48e7-bff4-0b873c97f5e2)
我们假设了观测中包含噪声，其中噪声服从正态分布。噪声正态分布如下式:
![image](https://github.com/user-attachments/assets/530b55a0-aed0-4362-bf84-5cd6c7d057a7)
其中，ϵ∼N(0,σ2)。
因此，我们现在可以写出通过给定的x观测到特定y的似然（likelihood）
![image](https://github.com/user-attachments/assets/ec44d347-4b18-4f3e-8066-6b54b7f46f07)
其实就是把均值换成了线性回归的表达式
根据极大似然估计法，参数w和b的最优值是使整个数据集的似然最大的值：
![image](https://github.com/user-attachments/assets/4f783b61-0126-41aa-af27-c09a084228cd)
极大似然估计：哪个高斯分布能让采样数据x出现的概率最大，由于每个数据点被采样的概率是相互独立的，故体现为概率的连乘，但是概率是小于一的，多个概率连乘会得到一个很小的数，很难计算，故现实中采取为最小化负对数似然−logP(y|X)。由此可以得到的数学公式是：
![image](https://github.com/user-attachments/assets/57f0337b-a867-48cc-887e-b27685db5a58)
在高斯噪声的假设下，最小化均方误差（最小二乘估计）等价于对线性模型的极大似然估计。

## 分类问题
一种表示分类数据的简单方法：独热编码（one‐hotencoding）。独热编码是一个向量，它的分量和类别一样多。类别对
应的分量设置为1，其他所有分量设置为0。

## Softmax的实现
避免计算exp(oj −max(ok))，而可以直接使用oj −max(ok)
![image](https://github.com/user-attachments/assets/4bc7be61-2ac3-48d6-b6fb-d145bf489e5a)
https://en.wikipedia.org/wiki/LogSumExp
## 激活函数
如果没有激活函数，那么单层网络和两层或多层网络实际上没有区别
 ReLU函数
![image](https://github.com/user-attachments/assets/176951d0-892a-41a4-9263-ce785a6ef149)
```python
y = torch.relu(x)
```
sigmoid函数
![image](https://github.com/user-attachments/assets/c4d2b7e4-ddbc-416e-8a01-9eb4d68c447b)
tanh函数
![image](https://github.com/user-attachments/assets/7bcc2065-d2c9-4334-bf22-a54f7067a0af)
## 训练误差和泛化误差
训练误差（trainingerror）是指，模型在训练数据集上计算得到的误差。泛化误差（generalizationerror）是指，模型应用在同样从原始样本的分布中抽取的无限多数据样本时，模型误差的期望。
## 独立同分布假设
对数据进行采样的过程没有进行“记忆”。换句话说，抽取的第2个样本和第3个样本的相关性，并不比抽取的第2个样本和第200万个样本的相关性更强。
## 欠拟合还是过拟合？
**一个重要因数是模型复杂度**
![image](https://github.com/user-attachments/assets/452196dc-b330-4597-8ab5-6e103bb76b31)

**另一个重要因素是数据集的大小。**训练数据集中的样本越少，我们就越有可能（且更严重地）过拟合。随着训练数据量的增加，泛化误差通常会减小。此外，一般来说，更多的数据不会有什么坏处。对于固定的任务和数据分布，模型复杂性和数据集大小之间通常存在关系。给出更多的数据，我们可能会尝试拟合一个更复杂的模型。能够拟合更复杂的模型可能是有益的。如果没有足够的数据，简单的模型可能更有用。对于许多任务，深度学习只有在有数千个训练样本时才优于线性模型。从一定程度上来说，深度学习目前的生机要归功于廉价存储、互联设备以及数字化经济带来的海量数据集。
## 范数与权重衰减
是我们应该如何精确地测量一个函数和零之间的距离呢？没有一个正确的答案。
**一种简单的方法是通过线性函数f(x)=w⊤x中的权重向量的某个范数来度量其复杂性**，要保证权重向量比较小，最常用方法是将其范数作为惩罚项加到最小化损失问题中，通常被称为岭回归（ridgeregression）
![image](https://github.com/user-attachments/assets/438ec76d-d7d2-494b-b7d7-134cfbc9ab60)
L1正则化线性回归，通常被称为套索回归（lassoregression）
L2正则化回归的小批量随机梯度下降更新如下式：
![image](https://github.com/user-attachments/assets/55a0e644-4547-48a6-8ec2-db515d7c0331)
与特征选择相比，权重衰减为我们提供了一种连续的机制来调整函数的复杂度。较小的λ值对应较少约束的w，而较大的λ值对w的约束更大。是否对相应的偏置b2进行惩罚在不同的实践中会有所不同，在神经网络的不同层中也会有所不同。通常，网络输出层的偏置项不会被正则化。
## 暂退法
**简单性的另一个角度是平滑性**。例如，当我们对图像进行分类时，我们预计向像素添加一些随机噪声应该是基本无影响的。1995年，克里斯托弗·毕晓普证明了具有输入噪声的训练等价于Tikhonov正则化(Bishop,1995)。这项工作用数学证实了“要求函数光滑”和“要求函数对输入的随机噪声具有适应性”之间的联系。然后在2014年，斯里瓦斯塔瓦等人(Srivastavaetal., 2014) 就如何将毕晓普的想法应用于网络的内部层提出了一个想法：在训练过程中，他们建议在计算后续层之前向网络的每一层注入噪声。因为当训练一个有多层的深层网络时，注入噪声只会在输入‐输出映射上增强平滑性。这个想法被称为暂退法（dropout）
## 前向传播
前向传播（forwardpropagation或forwardpass）指的是：按顺序（从输入层到输出层）计算和存储神经网络中每层的结果。
前向传播计算图
![image](https://github.com/user-attachments/assets/d065023b-a984-43e8-889e-30656a05eeec)
## 反向传播
反向传播（backwardpropagation或backpropagation）指的是计算神经网络参数梯度的方法。简言之，该方法根据微积分中的链式规则，按相反的顺序（沿着计算图反方向）从输出层到输入层遍历网络。回顾高等数学多元函数求偏导
## 训练神经网络
在训练神经网络时，前向传播和反向传播相互依赖。对于前向传播，我们沿着依赖的方向遍历计算图并计算其路径上的所有变量。然后将这些用于反向传播，其中计算顺序与计算图的相反。以上述简单网络为例：一方面，在前向传播期间计算正则项取决于模型参数W(1)和W(2)的当前值。它们是由优化算法根据最近迭代的反向传播给出的。另一方面，反向传播期间参数(4.7.11)的梯度计算，取决于由前向传播给出的隐藏变量h的当前值。因此，在训练神经网络时，在初始化模型参数后，我们交替使用前向传播和反向传播，利用反向传播给出的梯度来更新模型参数。注意，反向传播重复利用前向传播中存储的中间值，以避免重复计算。带来的影响之一是我们需要保留中间值，直到反向传播完成。这也是训练比单纯的预测需要更多的内存（显存）的原因之
一。此外，这些中间值的大小与网络层的数量和批量的大小大致成正比。因此，使用更大的批量来训练更深层次的网络更容易导致内存不足（outofmemory）错误。

## 梯度消失和梯度爆炸和对称
梯度消失和梯度爆炸，就是梯度无法收敛
而对称则是神经元采用了相同的权重，这样总体看来一层的神经元和一个神经元没有区别
## 参数初始化
解决（或至少减轻）梯度消失和梯度爆炸和对称问题的一种方法是进行参数初始化
默认初始化：我们使用正态分布来初始化权重值（框架默认）
Xavier初始化：没懂，也没必要，反正有用就对了
## 环境和分布偏移
通过将基于模型的决策引入环境，我们可能会破坏模型。（人总会和决策作博弈）
分布偏移的类型：
协变量偏移：虽然输入的分布可能随时间而改变，但标签函数（即条件分布P(y|x)）没有改变。统计学家称之为协变量偏移（covariateshift）通俗来说就是输入可能是柯基或是哈士奇或是卡通狗，但他们的标签都是狗
标签偏移：标签偏移（labelshift）描述了与协变量偏移相反的问题。这里我们假设标签边缘概率P(y)可以改变，但是类别条件分布P(x|y)在不同的领域之间保持不变。当我们认为y导致x时，标签偏移是一个合理的假设。例如，预测患者的疾病，我们可能根据症状来判断，即使疾病的相对流行率随着时间的推移而变化。
概念偏移：之前🐔就是🐔，现在🐔可能指蔡徐坤
## 经验风险与实际风险
经验风险（empiricalrisk）是为了近似真实风险（truerisk），整个训练数据上的平均损失，即从其
真实分布p(x,y)中抽取的所有数据的总体损失的期望值：
![image](https://github.com/user-attachments/assets/4f3d1577-9f9c-4fe2-8905-146e60b2adfe)
然而在实践中，我们通常无法获得总体数据。因此，经验风险最小化即在中最小化经验风险，是一种实用的机器学习策略，希望能近似最小化真实风险。
## 协变量偏移纠正
假设对于带标签的数据(xi,yi)，我们要评估P(y|x)。然而观测值xi是从某些源分布q(x)中得出的，而不是从
目标分布p(x)中得出的。幸运的是，依赖性假设意味着条件分布保持不变，即：p(y|x)=q(y |x)。
**怎么理解：这张图片的标签是狗是因为这张图片上画的是狗，跟生物圈里狗的分布没有关系**
如果源分布q(x)是“错误的”，我们可以通过在真实风险的计算中，使用以下简单的恒等式来进行纠正：
![image](https://github.com/user-attachments/assets/1eed76df-4865-454a-87b1-0e77225a8be9)
换句话说，我们需要根据数据来自正确分布与来自错误分布的概率之比，来重新衡量每个数据样本的权重：
![image](https://github.com/user-attachments/assets/8bf85816-d50b-4a6b-8861-0142fda5217c)
将权重βi代入到每个数据样本(xi,yi)中，我们可以使用”加权经验风险最小化“来训练模型：
![image](https://github.com/user-attachments/assets/53e76a3e-550e-47e5-9961-244204426a5b)
**那么如何知道这个权重β呢？书中给出的方法是用softmax回归训练一个2分类模型h(x)来获取β**
假设我们分别从p(x)和q(x)两个分布中抽取相同数量的样本。现在用z标签表示：从p抽取的数据为1，从q抽取的数据为−1。然后，混合数据集中的概率由下式给出：
![image](https://github.com/user-attachments/assets/1e8ecb0b-a22e-46a6-959a-6fd1cac7e5e9)
![image](https://github.com/user-attachments/assets/10d7f20e-9c70-4629-9af4-21beb367d0a4)
![image](https://github.com/user-attachments/assets/565abd91-3be4-44b6-9351-a68fd99e6f04)
以下是纠正协变量偏移的典型算法：
1. 生成一个二元分类训练集：{(x1,−1),...,(xn,−1),(u1,1),...,(um,1)}。
2. 用对数几率回归训练二元分类器得到函数h。
3. 使用βi =exp(h(xi))或更好的βi = min(exp(h(xi)),c)（c为常量）对训练数据进行加权。
4. 使用权重βi进行(4.9.5)中{(x1,y1),...,(xn,yn)}的训练。
